lib/elixir/lib/exception.ex

```diff
@@ -1569,6 +1569,13 @@ defmodule ErlangError do
     %FunctionClauseError{module: mod, function: fun, arity: arity}
   end
 
+  def normalize(:no_translation, stacktrace) do
```

Comment:
Question: Should this really be an `ArgumentError`?

Alternative:

* use `ErlangError`
* create a new `NoTranslationError` (open to better names), which has both encodings as structured fields

Reply:
It should be `ErlangError`. Otherwise we would slightly change the semantics of the code. Only badarg should become argument errors.

If this clause is not necessary in new Erlang versions, then we should remove it altogether. It is fine if the best error message only happens for those running on latest.

Reply:
I do not think removing the clause altogether would be a good option:

Formatted Error right now:

```
iex(1)> {:ok, fp} = :file.open('/tmp/file.uni', [:read, :write, {:encoding, :unicode}])
{:ok, #PID<0.120.0>}
iex(2)> :io.put_chars(fp, <<222>>)
** (ArgumentError) could not convert argument encoding:

  * 1st argument: device failed to transcode string from unicode to unicode

    (stdlib 3.17.1) io.erl:99: :io.put_chars(#PID<0.120.0>, <<222>>)
    iex:3: (file)
```

Formatted Error after removing the clause:

```
iex(1)> {:ok, fp} = :file.open('/tmp/file.uni', [:read, :write, {:encoding, :unicode}])
{:ok, #PID<0.115.0>}
iex(2)> :io.put_chars(fp, <<222>>)
** (ErlangError) Erlang error: :no_translation
    (stdlib 3.17.1) io.erl:99: :io.put_chars(#PID<0.115.0>, <<222>>)
    iex:2: (file)
```

Reply:
Please try removing the clause but formatting the error message in `def message/1` instead.

Reply:
For reference the error as reported in erlang itself:

```
1> {ok, Dev} = file:open('/tmp/file.unicode', [read, write, {encoding, unicode}]).
{ok,<0.87.0>}
2> io:put_chars(Dev, <<222>>).                                                    
** exception error: no_translation
     in function  io:put_chars/2
        called as io:put_chars(<0.87.0>,<<"Þ">>)
        *** argument 1: device failed to transcode string from unicode to unicode
```

Reply:
Actually, ignore me. It is not possible to format the error in `def message/1` because we no longer have the stacktrace.

Reply:
I know! We can add a :reason field to ErlangError and return the error_info there if any.

and then change message/1 to include the reason in the message, if any.

this way we can error_info for all ErlangErrors. No longer needing to special case :no_translation.

Reply:
@josevalim I'll try to achieve that then :+1:

Reply:
@josevalim The handling of the extra reason is a bit strange since we do not have a `message` field liek the other exceptions.

I see multiple ways on how to achieve our goal, what do you prefer?

## Option Field `reason`

```diff
diff --git a/lib/elixir/lib/exception.ex b/lib/elixir/lib/exception.ex
index d204b2a79..5dbb14fa3 100644
--- a/lib/elixir/lib/exception.ex
+++ b/lib/elixir/lib/exception.ex
@@ -1473,11 +1473,17 @@ def message(exception) do
 end
 
 defmodule ErlangError do
-  defexception [:original]
+  defexception [:original, :reason]
 
   @impl true
-  def message(exception) do
-    "Erlang error: #{inspect(exception.original)}"
+  def message(exception)
+
+  def message(%__MODULE__{original: original, reason: nil}) do
+    "Erlang error: #{inspect(original)}"
+  end
+
+  def message(%__MODULE__{reason: reason}) do
+    reason
   end
 
   @doc false
@@ -1569,19 +1575,15 @@ def normalize({:badarg, payload}, _stacktrace) do
     %ArgumentError{message: "argument error: #{inspect(payload)}"}
   end
 
-  def normalize(other, _stacktrace) do
-    %ErlangError{original: other}
+  def normalize(other, stacktrace) do
+    case error_info(other, stacktrace, "Erlang error: #{inspect(other)}") do
+      {:ok, message} -> %ErlangError{original: other, reason: message}
+      :error -> %ErlangError{original: other}
+    end
   end
 
   defp from_stacktrace([{module, function, args, _} | _]) when is_list(args) do
```

## Option Field `message`

```diff
diff --git a/lib/elixir/lib/exception.ex b/lib/elixir/lib/exception.ex
index d204b2a79..aa323d7ab 100644
--- a/lib/elixir/lib/exception.ex
+++ b/lib/elixir/lib/exception.ex
@@ -1473,12 +1473,7 @@ def message(exception) do
 end
 
 defmodule ErlangError do
-  defexception [:original]
-
-  @impl true
-  def message(exception) do
-    "Erlang error: #{inspect(exception.original)}"
-  end
+  defexception [:original, :message]
 
   @doc false
   def normalize(:badarg, stacktrace) do
@@ -1569,19 +1564,16 @@ def normalize({:badarg, payload}, _stacktrace) do
     %ArgumentError{message: "argument error: #{inspect(payload)}"}
   end
 
-  def normalize(other, _stacktrace) do
-    %ErlangError{original: other}
+  def normalize(other, stacktrace) do
+    default_message = "Erlang error: #{inspect(other)}"
+    case error_info(other, stacktrace, default_message) do
+      {:ok, message} -> %ErlangError{original: other, message: message}
+      :error -> %ErlangError{original: other, message: default_message}
+    end
   end
 
   defp from_stacktrace([{module, function, args, _} | _]) when is_list(args) do
```

## Option `error_info` inside `ErlangError`

```diff
diff --git a/lib/elixir/lib/exception.ex b/lib/elixir/lib/exception.ex
index d204b2a79..87893cec5 100644
--- a/lib/elixir/lib/exception.ex
+++ b/lib/elixir/lib/exception.ex
@@ -1473,11 +1473,19 @@ def message(exception) do
 end
 
 defmodule ErlangError do
-  defexception [:original]
-
-  @impl true
-  def message(exception) do
-    "Erlang error: #{inspect(exception.original)}"
+  defexception [:original, :reason]
+
+  @impl Exception
+  def message(exception)
+  def message(%__MODULE__{original: original, reason: nil}) do
+    "Erlang error: #{inspect(original)}"
+  end
+  def message(%__MODULE__{original: original, reason: reason}) do
+    default_message = "Erlang error: #{inspect(original)}"
+    case error_info(original, reason, default_message) do
+      {:ok, message} -> message
+      :error -> default_message
+    end
   end
 
   @doc false
@@ -1569,19 +1577,12 @@  def normalize({:badarg, payload}, _stacktrace) do
     %ArgumentError{message: "argument error: #{inspect(payload)}"}
   end
 
-  def normalize(other, _stacktrace) do
-    %ErlangError{original: other}
+  def normalize(other, stacktrace) do
+    %ErlangError{original: other, reason: stacktrace}
   end
 
   defp from_stacktrace([{module, function, args, _} | _]) when is_list(args) do
```

Reply:
We typically use :reason in cases like this. I want to avoid using message because the reason can be large and I want to avoid keeping the inspected structure in memory prematurely.

The only change I would make to the "reason" patch is to not change the message. I.e. I think we want to something like this:

```
Erlang error: :no_translation

  * 1st argument: device failed to transcode string from unicode to unicode
```

Reply:
`error_info` currently outputs the reason followed by a colon on the first line and can override it as well: `reason = Map.get(extra, :reason, default_reason)`

I can do it like you proposed, but then I would need to split the `error_info` into two functions (one that only handels the extra information and one that also handles the reason.

Is this the way to go?

PS: As soon as I have this all figured out, I will split the PR into two. One handling the `no_translation` and one changing the exceptions in general.

Reply:
@maennchen you can pass the reason to be an empty string and I think it will give us the desired result :)

Reply:
Can we change it to return {:ok, reason, details}? And for those cases we would ignore the reason?

Reply:
Ok, now I will open a PR in a few minutes with the new ErlangError details. This PR now only contains the aligned error handling to OTP.